### PR TITLE
Add multi-block support for matmul_2d

### PR DESCRIPTION
**Pull request by yugaoTT**
### Problem description
Current matmul 2d only supports single output block per core, we need to change it so that large matmuls can use less L1.

### What's changed
support multiple output blocks on both height and width for interleaved output. 
support only multi-block on height for sharded output. 
support in0 with multi-block on height.
support in1/bias with multi-block on width.

so we will have 3 levels of blocks:
per_core_M
   block_h
      subblock_h
previously we missed the second one

### Checklist
- [x] Post commit CI https://github.com/tenstorrent/tt-metal/actions/runs/11842667851/job/33003438280
- [x] Blackhole Post  https://github.com/tenstorrent/tt-metal/actions/runs/11846243832
- [x] Model regression https://github.com/tenstorrent/tt-metal/actions/runs/11845112861
- [x] Device performance https://github.com/tenstorrent/tt-metal/actions/runs/11845126185
- [x] t3k freq https://github.com/tenstorrent/tt-metal/actions/runs/11840255278
- [x] nightly https://github.com/tenstorrent/tt-metal/actions/runs/11845135114

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/matmul/device/matmul_op_multi_core_reuse_optimized_program_factory.cpp [hunk 1]
```diff
@@ -206,6 +206,8 @@ operation::ProgramWithCallbacks create_program(
         in1_per_core_w,       // in1_per_core_w
 
         num_blocks,  // num_blocks
+        1,
```
Comment (rfurko-tt) [inline ttnn/cpp/ttnn/operations/matmul/device/matmul_op_multi_core_reuse_optimized_program_factory.cpp:209]: could you please add comment to represent that it's output blocks number (information to make sure that it's clear what this one represents)?

File: ttnn/cpp/ttnn/operations/matmul/device/matmul_op_multi_core_reuse_mcast_2d_program_factory.cpp [hunk 2]
```diff
@@ -89,37 +91,52 @@ operation::ProgramWithCallbacks create_program_mcast_in0_in1(
     const bool in1_is_sharded = in1_buffer->buffer_layout() == TensorMemoryLayout::WIDTH_SHARDED;
     const bool output_is_sharded = out_buffer->buffer_layout() == TensorMemoryLayout::BLOCK_SHARDED;
 
-    uint32_t in0_block_tiles = per_core_M * in0_block_w;
+    bool do_not_inplace_interm0_out_CB = output_is_sharded && (per_core_M != out_block_h);
+
+    // hardcode testing, testing multiple blocks on height for now
+    uint32_t in0_block_h = out_block_h;
```
Comment (rfurko-tt) [inline ttnn/cpp/ttnn/operations/matmul/device/matmul_op_multi_core_reuse_mcast_2d_program_factory.cpp:97]: will you remove this code once ready to merge? question based on the comment

**Review by rfurko-tt — COMMENTED** [review]
Great work! Thanks for making matmul better!

1. Curious, what's the expected performance improvement? 
I expect that when we didn't fit in L1 we used empty configuration that didn't use L1, right? Please correct me if I am wrong.
2. What's the difference in performance if we use (4, 4) vs (8, 8) for num_out_blocks_x where both fit? Just curious how much tuning do we need to get the best performance yet avoid custom configuration for each layer. 

Thanks in advance!

**Comment by yugaoTT** [comment]
Hey @rfurko-tt , for 1 I won't expect perf difference when when everthing already fits L1. It's more about supporting the large matmuls where we cannot run currently (16k x 16k for example). with the change, now we can compute output block by block now, and can run those large matmuls. 

for 2. when the matmul both fits 4x8 and 8x8, and it's for a large matmul (compute bound) I think perf is doubled.

File: ttnn/cpp/ttnn/operations/matmul/device/matmul_op.cpp [hunk 3]
```diff
@@ -1183,7 +1189,12 @@ void Matmul::validate(
                 // tensor in1
                 TT_FATAL(input_tensor_b.memory_config().memory_layout == TensorMemoryLayout::WIDTH_SHARDED, "Error");
             } else if constexpr (std::is_same_v<ProgramConfigType, MatmulMultiCoreReuseMultiCastProgramConfig>) {
+                TT_FATAL(program_config.per_core_M % program_config.out_block_h == 0, "Error");
+                TT_FATAL(program_config.per_core_N % program_config.out_block_w == 0, "Error");
+                TT_FATAL(program_config.out_block_h % program_config.out_subblock_h == 0, "Error");
+                TT_FATAL(program_config.out_block_w % program_config.out_subblock_w == 0, "Error");
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/matmul/device/matmul_op.cpp:1195]: These should print out something like "Error: incompatible values {} and {}"

**Review by bbradelTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/matmul/device/matmul_op.cpp (hunk 3 above)
Comment (yugaoTT) [reply ttnn/cpp/ttnn/operations/matmul/device/matmul_op.cpp]: done

File: ttnn/cpp/ttnn/operations/matmul/device/matmul_op_multi_core_reuse_mcast_2d_program_factory.cpp (hunk 2 above)
Comment (yugaoTT) [reply ttnn/cpp/ttnn/operations/matmul/device/matmul_op_multi_core_reuse_mcast_2d_program_factory.cpp:96]: removed outdated comment

File: ttnn/cpp/ttnn/operations/matmul/device/matmul_op_multi_core_reuse_optimized_program_factory.cpp (hunk 1 above)
Comment (yugaoTT) [reply ttnn/cpp/ttnn/operations/matmul/device/matmul_op_multi_core_reuse_optimized_program_factory.cpp]: done

File: tests/tt_eager/integration_tests/test_bert.cpp [hunk 4]
```diff
@@ -33,6 +33,8 @@ Tensor encoder(Tensor&& hidden_states, const Tensor& attention_mask, const Param
         .in0_block_w = 4,
         .out_subblock_h = 4,
         .out_subblock_w = 2,
+        .out_block_h = 12,
+        .out_block_w = 8,
         .per_core_M = 12,
         .per_core_N = 8,
```
Comment (TT-BrianLiu) [inline tests/tt_eager/integration_tests/test_bert.cpp:39]: We should probably clean this up:
- `in0_block_w` can be renamed `inner_dim_block` (or something that is not specific to in0)
- Pack all the `*_h` and `*_w` together?
- `per_core_M/N` should be renamed to `per_core_h/w` to be in line with the other parameters

**Review by TT-BrianLiu — APPROVED** [review]

**Review by mywoodstock — APPROVED** [review]

File: tests/tt_eager/integration_tests/test_bert.cpp (hunk 4 above)
Comment (yugaoTT) [reply tests/tt_eager/integration_tests/test_bert.cpp:39]: issue for tracking
https://github.com/tenstorrent/tt-metal/issues/15054

File: ttnn/cpp/ttnn/operations/matmul/device/kernels/dataflow/reader_bmm_tile_layout_in1_receiver_writer_padding.cpp [hunk 5]
```diff
@@ -91,91 +100,113 @@ void kernel_main() {
...
+                    cb_push_back(cb_id_in1, in1_block_num_tiles);
+                }
 
 #ifdef FUSE_BIAS
-        // Only read bias on first batch
-        if (b == 0) {
-            // Operand 2
-            cb_reserve_back(cb_id_in3, in3_block_w);
+                // Only read bias on first batch, or we have multiple output blocks
+                if ((b == 0 && bh == 0) || num_blocks_w_dim > 1) {
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/matmul/device/kernels/dataflow/reader_bmm_tile_layout_in1_receiver_writer_padding.cpp:126]: Note after this PR is merged: I wonder if the extra conditions may be causing issues in some scenarios.

Running into a hang related to these lines of code. See https://github.com/tenstorrent/tt-metal/issues/15732#issuecomment-2566682486

**Pull request merged**